### ISSUE TITLE
Use sensible end_time when editing events without it

### ISSUE THREAD

**Issue by igal** [calagator/calagator#16]
Currently this defaults to fill these fields with the current time, which is unreasonable.

One solution is to leave these fields blank, to match the fact that they have no data in them. It should display a warning about the end time not being set.

Another solution is to prefill these with the current time + 2 hours, because this makes it easier to actually set the time. It should display a warning about the end time being prefilled.

**Comment by jwithington** [comment]
This seems to already have been addressed, as creating a new event adds an end time of an hour later. Yeah?

**Comment by donaldedgar** [comment]
The end time seems to be fine. But if you create a new event in the future without an end time, and then edit it, it will suggest today's date for the end date, which is really only helpful if we're running backward in time.

**Comment by jwithington** [comment]
A ha!  I misunderstood the issue. Carry on. :smile:

**Comment by reidab** [comment]
If anyone's trying to repro this issue, importing http://www.meetup.com/PDX-Creative-Coders/events/219455378/ will result in the creation of an event with no end time.

**Comment by sipple** [comment]
I'm looking into this bug now, and have some questions as to how you'd like this handled.

Currently, if not present, both start_time and end_time are calculated off of Time.now. A simple change to keep end_time from being based off of Time.now in cases where start_time is present is to calculate start_time off of Time.now, but use start_time + 1 hour to calculate end time. 

This covers both the new event scenario that the current method handles, and providing an in-the-future default end_time when editing an event that has a start_time but no end_time.

So, questions:

* Is that the direction you'd prefer, or would you rather this be left blank? If blank is the preferred direction, this should be pre-populated.

* Whichever direction we go, a warning needs to be displayed. I'm new to the project, and am not sure what the preferred way of displaying a warning would be. I see that field error messages are displayed using semantic_errors in cases where a record fails validation on save, but adding an error to the field to get it to display there feels wrong (but maybe not!). If we went that direction, the logic would need to be that end_time isn't present and that it's the edit action, not new.

* Do you have a preference on what the warning message wording should be?

**Comment by shawnacscott** [comment]
Hi @sipple, welcome!

Personally, I'd like to see the field left blank if there's no specified end time. Events are allowed to have blank end times, so that seems most sensible to me. 

For the error/warning message, I'd like to see a red box (similar to the box shown when an input gets focus) surrounding the input with a subscript of the same color (similar to the helper text on the "Venue" input field) that says something like, "No end time detected", or "End time blank by default". I'm not in love with either of those, though, so feel free to take some ownership of that. :smile_cat:

**Comment by reidab** [comment]
:+1: I agree that leaving the field blank when editing events with no end time is perfectly reasonable. 

For new events, there's also some room for flexibility here. People rarely create new events that start _right now_, so the pre-population that we do doesn't necessarily make sense. Watching the change log, I've definitely seen people add something with the current time as its start time and then have to go edit it to reflect the real start time.

I'd propose:

* remove the pre-population of both start and end times for new events
* make sure the calendar widget still defaults to the current month on activation for blank fields
* keep the JS logic that updates the end time when the start time is changed

**Comment by sipple** [comment]
Thank you @shawnacscott and @reidab!

If we're not going to pre-populate the end date when either creating a new event or editing one, should we display a warning about the end time at all? It sounds like a start time with no end time is a valid use case, but I could be wrong. If that is a valid use case, I think leaving it blank by default in all cases makes sense.

If we want to leave start date blank when creating a new event, though, we'll need to call that field out as required, as Event Title is. This could be a big confusing, though, since it's part of the same line where End Date/Time is listed, and it appears the current method for indicating required is a "*" beside the name.

It might make sense to continue to pre-populate the start date and time for now (as the simpler immediate fix), but implement javascript validation across the board in the future for required fields?

I apologize if I'm overcomplicating a simple issue for y'all.

**Comment by shawnacscott** [comment]
:+1: everything from @reidab. In this case, I think no warning makes sense.

@sipple In my opinion, the easiest solution to whether or not to call out something as required is to simply break end time to a new line like for mobile views and just make start time as required. The labels could change to be "Start time"/"End time" rather than "When".

Alternatively, I'm fine with just marking the field required and not *actually* requiring an end time, but. That may be just me.

**Comment by reidab** [comment]
I like the idea of keeping the whole field on one line, marking it as required, but not actually requiring an end time on the backend. 

The only reason that we don't require an end time is that we support importing from sources that don't always include one. When people are adding events manually, we want prod them into adding an end time, since it lets us provide more accurate information about the event. This is why we auto-populate it when changing the start time. There does exist a small class of events with no pre-set end time, but I think it's fine to make people manually clear the end time fields if they really want to create one.

**Comment by sipple** [comment]
:+1: Sounds good to me. I'm mostly done with this.

One snag: the required field asterisk is inserted by simple_form, which detects if an input is required and adds a <abbr> tag in with the asterisk accordingly when it renders that label's input. For "When", it's just a straight label, and from what I can tell, simple_form can't be told a label is "required" and render it accordingly. If I'm wrong about that, please let me know.

If that is the case, most of the solutions I can think of are a bit hacky. The best of which, I think, might be to change the "When" label and manually put in the <abbr> tag. That would look like this:
```erb
<%= f.label :when, :class => 'label'  do -%>
  When<abbr title="required">*</abbr>
<% end -%>
```

Also: From the testing I've done on calagator.org and on my dev system, the javascript only seems to update the end date on start time changes when there's already a date and time in the end date. If we start with an empty start_date, that javascript will need to be changed. Is that something you'd want done as part of this issue, or would it make sense as a separate one?

**Comment by reidab** [comment]
Adding the required HTML manually, while a bit hacky, does indeed seem like the best option.

If you're willing to take a look at the javascript changes, I think they'd be appropriate within the work here.

**Comment by reidab** [comment]
Fixed by #428!
